In [1]:
Output = '/Users/alexis/Library/CloudStorage/OneDrive-UniversityofNorthCarolinaatChapelHill/CEMALB_DataAnalysisPM/Projects/P1015. Fire Sufficient Similarity/P1015.3. Analyses/P1015.3.2. Data Processing/Output'
cur_date = '030525'

library(missForest)
library(readxl)
library(openxlsx)
library(tidyverse)
library(imputeLCMD)
library(factoextra)

# reading in files
ws_df = data.frame(Data = 'WS', read_excel("Input/Woodsmoke_Data_012825.xlsx", sheet = 2)) %>%
    select(-Sample_Number)
wf_df = data.frame(Data = 'WF', read_excel("Input/Wildfire_Data_012825.xlsx", sheet = 2)) %>%
    select(-Sample_Number)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.3     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Loading required package: tmvtnorm

Loading required package: mvtnorm

Loading required package: Matrix


Attaching package: ‘Matrix’


The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack


Loading required package: stats4

Loading required package: gmm

Loading required package: sandwich

Loading required package: norm

This package has some major limitations
(for example, it does not work reliably when
the number of variables exceeds 30),
an

NOTES TO SELF:
- change 'replicate' to 'instance' and 'chemical class' to 'class'

In [2]:
head(ws_df)
head(wf_df)

Data HAWC_ID Study                    Replicate Chemical_Class Metric
1 WS    821855 Erlandsson et al. 2020    1        PAH            Weight
2 WS    821855 Erlandsson et al. 2020    1        PAH            Weight
3 WS   1257056 McCarrick et al. 2024     1        PAH            Weight
4 WS    267140 Alfheim and Ramdahl 1984  1        PAH            Volume
5 WS   1263480 Burnet et al. 1990        1        PAH            Volume
6 WS   1263480 Burnet et al. 1990       NA        PAH            Volume
  DTXSID        Name               Value             
1 DTXSID3074787 1-Methylanthracene 11.3              
2 DTXSID3074787 1-Methylanthracene 2.2999999999999998
3 DTXSID3074787 1-Methylanthracene 1.66              
4 DTXSID3074787 1-Methylanthracene NA                
5 DTXSID3074787 1-Methylanthracene NA                
6 DTXSID3074787 1-Methylanthracene NA

Data HAWC_ID Study                Replicate Chemical_Class Metric
1 WF   1289821 Liang et al. 2021     1        PAH            Volume
2 WF   1289821 Liang et al. 2021     1        PAH            Volume
3 WF   1289737 Campbell et al. 2024  1        PAH            Weight
4 WF   1289739 Campos et al. 2019    1        PAH            Weight
5 WF   1289739 Campos et al. 2019   NA        PAH            Weight
6 WF   1289739 Campos et al. 2019   NA        PAH            Weight
  DTXSID         Name                                Value
1 NA             1-(10-methylanthracen-9-yl)ethanone 5    
2 DTXSID50176885 1-Acenaphthenone                    1    
3 DTXSID1074759  1-Methylchrysene                    NA   
4 DTXSID1074759  1-Methylchrysene                    NA   
5 DTXSID1074759  1-Methylchrysene                    NA   
6 DTXSID1074759  1-Methylchrysene                    NA

In [3]:
dim(ws_df)
dim(wf_df)

[1] 6607    9

[1] 4340    9

# Filter 1

Removing duplicate records for instances that weren't measured at all (MAR). 

In [4]:
`%notin%` <- Negate(`%in%`)

ws_df = ws_df %>%
    filter(Replicate %notin% NA)

wf_df = wf_df %>%
    filter(Replicate %notin% NA)

In [5]:
dim(ws_df)
dim(wf_df)

[1] 4203    9

[1] 2378    9

In [6]:
# combining dfs
combined_df = rbind(ws_df, wf_df)
head(combined_df)

Data HAWC_ID Study                    Replicate Chemical_Class Metric
1 WS    821855 Erlandsson et al. 2020   1         PAH            Weight
2 WS    821855 Erlandsson et al. 2020   1         PAH            Weight
3 WS   1257056 McCarrick et al. 2024    1         PAH            Weight
4 WS    267140 Alfheim and Ramdahl 1984 1         PAH            Volume
5 WS   1263480 Burnet et al. 1990       1         PAH            Volume
6 WS    267091 Forchhammer et al. 2012  1         PAH            Volume
  DTXSID        Name               Value             
1 DTXSID3074787 1-Methylanthracene 11.3              
2 DTXSID3074787 1-Methylanthracene 2.2999999999999998
3 DTXSID3074787 1-Methylanthracene 1.66              
4 DTXSID3074787 1-Methylanthracene NA                
5 DTXSID3074787 1-Methylanthracene NA                
6 DTXSID3074787 1-Methylanthracene NA

In [7]:
length(unique(combined_df$Name))

[1] 192

192 unique variables.

# Chemical/Metal Background Filters

Determining what variables have missing data for all variables delineated based on the dfs the data will be split into and analyzed. 

A variable (ie. `Name` which represents a chemical or metal) will be retained if >= 50% of its experimental data points were measurable values. In other words, chemicals were excluded if they were NAs (not evaluated) or had non-detect values across > 50% of the samples.

In [8]:
sample_type_presence_df = combined_df %>%
    # if an experimental value is there count it as being "present"
    mutate(count = ifelse(Value %notin% c('NA', 'ND'), 1, 0)) %>%
    # determining which have at least one experimental value within each sample type
    group_by(HAWC_ID, Metric, Name) %>%
    # calculating the percentage of experimental data points
    reframe(Variable_Presence_Percentage = (sum(count)/n()) * 100) %>%
    filter(Variable_Presence_Percentage >= 50)

head(sample_type_presence_df)

HAWC_ID Metric Name                 Variable_Presence_Percentage
1 266980  Volume Acetaldehyde         100                         
2 266980  Volume Benzene              100                         
3 266980  Volume Benzo(GHI)perylene   100                         
4 266980  Volume Benzo(a)pyrene       100                         
5 266980  Volume Benzo(b)fluoranthene 100                         
6 266980  Volume Benzo(k)fluoranthene 100

In [9]:
# filtering the original df
filter1_df = inner_join(sample_type_presence_df[,1:3], combined_df) %>%
    # metals don't have a DTXSID, so making that col their name
    mutate(DTXSID = ifelse(DTXSID != 'NA', DTXSID, Name)) 

head(filter1_df)

Joining with `by = join_by(HAWC_ID, Metric, Name)`


HAWC_ID Metric Name               Data Study                 Replicate
1 266980  Volume Acetaldehyde       WS   Stockfelt et al. 2012 1        
2 266980  Volume Acetaldehyde       WS   Stockfelt et al. 2012 2        
3 266980  Volume Benzene            WS   Stockfelt et al. 2012 1        
4 266980  Volume Benzene            WS   Stockfelt et al. 2012 2        
5 266980  Volume Benzo(GHI)perylene WS   Stockfelt et al. 2012 1        
6 266980  Volume Benzo(GHI)perylene WS   Stockfelt et al. 2012 2        
  Chemical_Class DTXSID        Value
1 VOC            DTXSID5039224 71000
2 VOC            DTXSID5039224 37000
3 VOC            DTXSID3039242 33000
4 VOC            DTXSID3039242 21000
5 PAH            DTXSID5023908 41   
6 PAH            DTXSID5023908 11

In [10]:
dim(combined_df)
dim(filter1_df)

[1] 6581    9

[1] 4189    9

Started with 6581 records, 2392 were removed, leaving 4189. 

# Second Chemical/Metal Background Filter

In the intial filter, WS and WF samples were combined. However, this time variables (ie. metal or chemical) will be split based on their `Metric` (ie. volume or weight) and then retained if that variable is found in both woodsmoke and wildfire samples within volume or weight samples.

In [11]:
split_filtered_df = filter1_df %>%
    group_by(Data, Metric) %>%
    group_split

split_ws_vol_df = split_filtered_df[[3]]
split_ws_weight_df = split_filtered_df[[4]]
split_wf_vol_df = split_filtered_df[[1]]
split_wf_weight_df = split_filtered_df[[2]]

In [12]:
# seeing how many unique variables are in each df and if they're consistent in each file
length(unique(split_ws_vol_df$Name))
length(unique(split_wf_vol_df$Name))
length(unique(split_ws_weight_df$Name))
length(unique(split_wf_weight_df$Name))

[1] 98

[1] 68

[1] 94

[1] 70

In [13]:
# they're not so first getting variables that are in weight or volume samples
consistent_wf_vol_df = split_wf_vol_df %>%
    filter(Name %in% unique(split_ws_vol_df$Name))
consistent_ws_vol_df = split_ws_vol_df %>%
    filter(Name %in% consistent_wf_vol_df$Name)
consistent_wf_weight_df = split_wf_weight_df %>%
    filter(Name %in% unique(split_ws_weight_df$Name))
consistent_ws_weight_df = split_ws_weight_df %>%
    filter(Name %in% consistent_wf_weight_df$Name)

length(unique(consistent_wf_vol_df$Name))
length(unique(consistent_ws_vol_df$Name))
length(unique(consistent_wf_weight_df$Name))
length(unique(consistent_ws_weight_df$Name))

[1] 33

[1] 33

[1] 34

[1] 34

There were 98, 68, 94 and 70 woodsmoke weight, woodsmoke volume, wildfire weight, and wildfire volume samples, respectively. 33 variables were common between volume samples and 34 were common between weight records and will be retained.

In [14]:
# recombining data
vol_df = rbind(consistent_ws_vol_df, consistent_wf_vol_df)
weight_df = rbind(consistent_ws_weight_df, consistent_wf_weight_df)

head(vol_df)

HAWC_ID Metric Name                 Data Study                 Replicate
1 266980  Volume Benzene              WS   Stockfelt et al. 2012 1        
2 266980  Volume Benzene              WS   Stockfelt et al. 2012 2        
3 266980  Volume Benzo(GHI)perylene   WS   Stockfelt et al. 2012 1        
4 266980  Volume Benzo(GHI)perylene   WS   Stockfelt et al. 2012 2        
5 266980  Volume Benzo(k)fluoranthene WS   Stockfelt et al. 2012 1        
6 266980  Volume Benzo(k)fluoranthene WS   Stockfelt et al. 2012 2        
  Chemical_Class DTXSID        Value
1 VOC            DTXSID3039242 33000
2 VOC            DTXSID3039242 21000
3 PAH            DTXSID5023908 41   
4 PAH            DTXSID5023908 11   
5 PAH            DTXSID0023909 23   
6 PAH            DTXSID0023909 4.5

# QRILC Imputation

Imputing non-detect data using QRILC didn't work, so I'm just replacing the non-detect data with zeros.

In [15]:
# determining number of ND and NA values in each df
vol_df %>%
    filter(Value %in% c('NA', 'ND')) %>%
    group_by(Value) %>%
    summarize(n())

weight_df %>%
    filter(Value %in% c('NA', 'ND')) %>%
    group_by(Value) %>%
    summarize(n())

Value n()

Value n()
1 NA     2 
2 ND    34

In [16]:
dim(vol_df)
dim(weight_df)

[1] 744   9

[1] 1411    9

Only the weight dataframe had non-detect values representing MNAR data. The remaining MAR data from the weight df will be imputed using random forest (RF).

In [17]:
mar_weight_df = weight_df %>%
    filter(Value == 'NA')

preimputed_df = anti_join(weight_df, mar_weight_df) 
preimputed_df$Value = as.numeric(preimputed_df$Value)
mar_weight_df$Value = as.numeric(mar_weight_df$Value)

# reordering
preimputed_df = preimputed_df[,c(1,2,4:8,3,9)]

head(preimputed_df)

Joining with `by = join_by(HAWC_ID, Metric, Name, Data, Study, Replicate,
Chemical_Class, DTXSID, Value)`
Warning message:
“NAs introduced by coercion”
Warning message:
“NAs introduced by coercion”


HAWC_ID Metric Data Study                 Replicate Chemical_Class
1 267105  Weight WS   Danielsen et al. 2011 1         PAH           
2 267105  Weight WS   Danielsen et al. 2011 2         PAH           
3 267105  Weight WS   Danielsen et al. 2011 1         Metal         
4 267105  Weight WS   Danielsen et al. 2011 2         Metal         
5 267105  Weight WS   Danielsen et al. 2011 1         Metal         
6 267105  Weight WS   Danielsen et al. 2011 2         Metal         
  DTXSID        Name                 Value
1 DTXSID4020878 2-Methylphenanthrene 10.0 
2 DTXSID4020878 2-Methylphenanthrene 37.0 
3 Cr            Cr                   21.4 
4 Cr            Cr                   21.3 
5 Cu            Cu                   65.6 
6 Cu            Cu                   34.4

Gave up on QRILC imputation for all mnar data, since I didn't have enough instances to impute within each study. I also couldn't impute across the entire dataset given that the number of data instances for each study wasn't the same. Therefore, sample groups were assigned a value of 0 if they had two or fewer instances.

In [18]:
# creating a df for mnar data
mnar_df = preimputed_df %>%
    filter(is.na(Value)) 

mnar_df[c(9:12,15,22:27,30:34),9] = 0
head(mnar_df)

HAWC_ID Metric Data Study              Replicate Chemical_Class DTXSID       
1 271638  Weight WS   Jalava et al. 2012 4         PAH            DTXSID3021774
2 271638  Weight WS   Jalava et al. 2012 6         PAH            DTXSID3021774
3 271638  Weight WS   Jalava et al. 2012 7         PAH            DTXSID3021774
4 271638  Weight WS   Jalava et al. 2012 6         PAH            DTXSID5047740
5 271769  Weight WS   Uski et al. 2012   4         PAH            DTXSID3021774
6 271769  Weight WS   Uski et al. 2012   6         PAH            DTXSID3021774
  Name         Value
1 Acenaphthene NA   
2 Acenaphthene NA   
3 Acenaphthene NA   
4 Coronene     NA   
5 Acenaphthene NA   
6 Acenaphthene NA

In [19]:
QRILC_imputation = function(dataset){
    set.seed(12)
      # """
    # Creating a quantile normalization function to normalize each sample.
    # :param (input): exposed and unexposed (vehicle) dfs
    # :output: 1 quantile normalized df
    # """
    #ADD WORDS
    wider_dataset = dataset %>% 
        # removing these cols temporarily 
        select(-c("Chemical_Class", "Name")) %>%
        pivot_wider(names_from = DTXSID, values_from = Value)

    # normalizing data since that what the QRILC function wants
    # had to pseudo log transform to prevent Inf values
    QRILC_prep = wider_dataset[,6:ncol(wider_dataset)] %>%
         mutate_all(., function(x) log10(x + 1)) %>%
         as.matrix()
                        
    imputed_QRILC_object = impute.QRILC(QRILC_prep, tune.sigma = 0.1)
    QRILC_log10_df = data.frame(imputed_QRILC_object[1]) 
    
    # converting back the original scale
    QRILC_df = QRILC_log10_df %>%
        mutate_all(., function(x) 10^x - 1)
    
    imputed_dataset = data.frame(cbind(unique(dataset[,1:5]), QRILC_df)) %>%
         pivot_longer(cols = 6:ncol(wider_dataset), names_to = "DTXSID", values_to = "Value")         

    return(imputed_dataset)

}

In [20]:
# will impute within each study
df_for_ids = mnar_df[is.na(mnar_df$Value),]
study_ids = unique(df_for_ids$HAWC_ID)

imputed_df = data.frame()
for (i in 1:length(study_ids)){

    filtered_df = preimputed_df %>%
        filter(HAWC_ID == study_ids[i])

    # calling fn
    imp_df = QRILC_imputation(filtered_df)
    imputed_df = rbind(imputed_df, imp_df)
    
}
# adding back in additional cols
imputed_df = inner_join(unique(imputed_df), unique(weight_df[,1:8]))
imputed_df = imputed_df[,c(1:5,9,6,8,7)]

head(imputed_df)

Joining with `by = join_by(HAWC_ID, Metric, Data, Study, Replicate, DTXSID)`


HAWC_ID Metric Data Study              Replicate Chemical_Class DTXSID       
1 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID3021774
2 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID4075459
3 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID8052691
4 271638  Weight WS   Jalava et al. 2012 1         Metal          Ca           
5 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID5047740
6 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID6062591
  Name                 Value 
1 Acenaphthene          132.7
2 Benzo(c)phenanthrene  493.9
3 Benzo(j)fluoranthene  617.3
4 Ca                   1140.0
5 Coronene               75.5
6 Cyclopenta(cd)pyrene 1067.0

In [21]:
imputed_weight_df = rbind(imputed_df, 
                          # adding back in non-imputed data
                          preimputed_df %>%
                              filter(HAWC_ID %notin% unique(mnar_df$HAWC_ID)),
                          # adding back in mar data
                          mar_weight_df, 
                          # adding back in mnar data that was converted to zeros
                          mnar_df[!is.na(mnar_df$Value),])

head(imputed_weight_df)
dim(weight_df)
dim(imputed_weight_df)

HAWC_ID Metric Data Study              Replicate Chemical_Class DTXSID       
1 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID3021774
2 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID4075459
3 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID8052691
4 271638  Weight WS   Jalava et al. 2012 1         Metal          Ca           
5 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID5047740
6 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID6062591
  Name                 Value 
1 Acenaphthene          132.7
2 Benzo(c)phenanthrene  493.9
3 Benzo(j)fluoranthene  617.3
4 Ca                   1140.0
5 Coronene               75.5
6 Cyclopenta(cd)pyrene 1067.0

[1] 1411    9

[1] 1248    9

In [22]:
# length(unique(imputed_df$HAWC_ID))
# length(unique(mar_weight_df$HAWC_ID))
# bleh = mnar_df[!is.na(mnar_df$Value),]
# length(unique(bleh$HAWC_ID))    
# bleh2 = drop_na(preimputed_df) %>%
#                                filter(HAWC_ID %notin% unique(mnar_df$HAWC_ID))
# length(unique(bleh2$HAWC_ID))    
# length(unique(weight_df$HAWC_ID))

# x = c(unique(imputed_df$HAWC_ID), unique(mar_weight_df$HAWC_ID),unique(bleh$HAWC_ID), unique(bleh2$HAWC_ID))
# x[duplicated(x)]

# weight_df %>%
#     filter(HAWC_ID %in% x[duplicated(x)][1])

In [23]:
# dim(imputed_df)
# dim(mar_weight_df)
# dim(mnar_df[!is.na(mnar_df$Value),])
# dim(drop_na(preimputed_df) %>%
#                               filter(HAWC_ID %notin% unique(mnar_df$HAWC_ID)))

In [24]:
# avg_df = preimputed_df %>%
#     group_by(HAWC_ID, Name) %>%
#     mutate(Avg_Value = mean(Value)) %>%
#     select(-c("Value", "Replicate")) %>%
#     unique()

# head(avg_df)

# RF Imputation

In [25]:
random_forest_imputation = function(dataset){
    # ADD WORDS
    wider_dataset = dataset %>%
        # removing these cols temporarily 
        select(-c("Chemical_Class", "Name")) %>%
        pivot_wider(names_from = DTXSID, values_from = Value)
    
    preimputed_df = wider_dataset[,6:ncol(wider_dataset)]
    
    imputed_RF_object = missForest(as.matrix(preimputed_df))
    imputed_RF_df = imputed_RF_object$ximp
    imputed_final_df = data.frame(cbind(wider_dataset[,1:5], imputed_RF_df)) %>%
        pivot_longer(cols = 6:ncol(wider_dataset), names_to = "DTXSID", values_to = "Value")
    
    return(imputed_final_df)
}

# calling fn
rf_imputed_df = random_forest_imputation(imputed_weight_df %>% 
                             filter(HAWC_ID == unique(mar_weight_df$HAWC_ID)))
# adding back in additional cols
rf_imputed_df = inner_join(rf_imputed_df, unique(weight_df[,1:8]))
rf_imputed_df = rf_imputed_df[,c(1:5,9,6,8,7)]

head(rf_imputed_df)

Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response h

HAWC_ID Metric Data Study                 Replicate Chemical_Class DTXSID
1 279220  Weight WS   Kasurinen et al. 2017 1         Metal          Ca    
2 279220  Weight WS   Kasurinen et al. 2017 1         Metal          Cr    
3 279220  Weight WS   Kasurinen et al. 2017 1         Metal          Cu    
4 279220  Weight WS   Kasurinen et al. 2017 1         Metal          Fe    
5 279220  Weight WS   Kasurinen et al. 2017 1         Metal          K     
6 279220  Weight WS   Kasurinen et al. 2017 1         Metal          Mn    
  Name Value  
1 Ca     980.0
2 Cr      74.0
3 Cu     360.0
4 Fe     340.0
5 K    12000.0
6 Mn     275.9

In [26]:
# combining data
final_df = full_join(imputed_weight_df %>% filter(HAWC_ID != unique(mar_weight_df$HAWC_ID)),
                            rf_imputed_df)

head(final_df)
dim(final_df)

Joining with `by = join_by(HAWC_ID, Metric, Data, Study, Replicate,
Chemical_Class, DTXSID, Name, Value)`


HAWC_ID Metric Data Study              Replicate Chemical_Class DTXSID       
1 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID3021774
2 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID4075459
3 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID8052691
4 271638  Weight WS   Jalava et al. 2012 1         Metal          Ca           
5 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID5047740
6 271638  Weight WS   Jalava et al. 2012 1         PAH            DTXSID6062591
  Name                 Value 
1 Acenaphthene          132.7
2 Benzo(c)phenanthrene  493.9
3 Benzo(j)fluoranthene  617.3
4 Ca                   1140.0
5 Coronene               75.5
6 Cyclopenta(cd)pyrene 1067.0

[1] 1248    9

In [27]:
# exporting
write.xlsx(final_df, paste0(Output,"/", "Processed_Weight_Data_", cur_date, ".xlsx"), 
           rowNames = FALSE)
write.xlsx(vol_df, paste0(Output,"/", "Processed_Volume_Data_", cur_date, ".xlsx"), 
           rowNames = FALSE)